v23.1/transaction-retry-error-reference.md

```diff
@@ -1,54 +1,185 @@
+                        if retry == max_retries:
+                            cur.execute("SET inject_retry_errors_enabled = 'false'")
+                        cur.execute("SELECT now()")
+                        logging.debug("status message: %s", cur.statusmessage)
+                else:
+                    op(conn)
+                # If we reach this point, we were able to commit, so we break
+                # from the retry loop.
+                return
+
+            except SerializationFailure as e:
+                # This is a retry error, so we roll back the current
+                # transaction and sleep for a bit before retrying. The
+                # sleep time increases for each failed transaction.
+                logging.debug("got error: %s", e)
+                conn.rollback()
+                sleep_ms = (2 ** retry) * 0.1 * (random.random() + 0.5)
+                logging.debug("Sleeping %s seconds", sleep_ms)
+                time.sleep(sleep_ms)
+
+            except psycopg2.Error as e:
+                logging.debug("got error: %s", e)
+                raise e
+
+        raise ValueError(f"Transaction did not succeed after {max_retries} retries")
+~~~
+
+Calling `run_transaction` without an `op` input sets `inject_retry_errors_enabled` as `true` until the final retry attempt, before which the `inject_retry_errors_enabled` is set back to `false`. For all attempts except the last one, CockroachDB will inject a retryable serialization error for the client to handle. If the client cannot handle the error properly, the retry logic isn't working properly.
+
+### Minimize transaction retry errors
+
+While the steps described in [Client-side retry handling](#client-side-retry-handling) detail how to configure your application to restart a failed transaction, there are also a number of changes you can make to your application logic to prevent transaction retry errors from occurring in the first place:
+
+1. Design your schema and queries to reduce contention. For more information about how contention occurs and how to avoid it, see [Transaction Contention](performance-best-practices-overview.html#transaction-contention). In particular, if you are able to [send all of the statements in your transaction in a single batch](transactions.html#batched-statements), CockroachDB can usually automatically retry the entire transaction for you.
+
+1. Use historical reads with [`SELECT ... AS OF SYSTEM TIME`](as-of-system-time.html).
+
+1. In some cases, [`SELECT FOR UPDATE`](select-for-update.html) can also be used to reduce the occurrence of serialization errors.
+
+1. [High priority transactions](transactions.html#transaction-priorities) are less likely to experience serialization errors than low priority transactions. Adjusting transaction priorities usually does not change how many serialization errors occur, but it can help control which transactions experience them.
+
```

Comment:
Same comment before your latest edit. Just swapped order to prioritize column family mitigation over closed timestamp interval

Reply:
Hey Andrew, just peeking on these docs as I start my work... should we mention the actual name of the setting here? It links to the doc I'm working on, but the setting name is mentioned there either. In docs, it was actually hard to find `kv.closed_timestamp.target_duration` as it is not public?

Reply:
I suppose this setting can have adverse effects on a bunch of things, so that is probably why it is not public. Given we mentioning it in docs though, do we need to think about it's status?

Reply:
Agreed, I think we should definitely name it here, if it's part of official guidance (the intent of this page). How else would users be able to implement it if we don't name it!! I'll make that update, thanks much, great call!